### PR TITLE
Honeycomb flakes

### PR DESCRIPTION
**Pull request by pfebrer**
Could a honeycomb/graphene flake geometry be useful? I'm using one currently and I think they are useful in general to create finite bilayers with arbitrary twist.

This PR is a proposal to add them.

**Examples**

```python
import sisl
import sisl.viz

sisl.geom.graphene_flake(shells=3).plot(axes="xy")
```

newplot (4)

```python
import sisl
import sisl.viz

sisl.geom.honeycomb_flake(shells=3, bond=1.6, atoms=["B", "N"]).plot(axes="xy")
```

newplot (5)

 - [ ] Closes #xxxx
 - [x] Tests added
 - [x] Ranned `isort .` at top--level
 - [x] Documentation for functionality, `docs/`
 - [x] Changes documented, `CHANGELOG.md`

### CODE REVIEW

**Comment by zerothi** [comment]
How about an API of `graphene_flakes` that resembles the `heteroribbon`? In that way attachments of various segments could be done?

**Comment by pfebrer** [comment]
What would you attach to what in this case? :sweat_smile:

**Comment by zerothi** [comment]
> What would you attach to what in this case? 😅

A new flake/ribbon, connecting ribbons and flakes could be quite interesting ;)

**Comment by pfebrer** [comment]
Hmm yes, but the construction of the flake itself is not a composite geometry, no?

So I would imagine you can take `graphene_flake` and implement a section that knows how to connect to ribbons, which determines the arguments for `graphene_flake` and maybe rotates it, etc...

**Comment by zerothi** [comment]
yes, that sounds cool :)
Ok, but anyways, I think this is fitting

File: pyproject.toml [hunk 1]
```diff
@@ -11,7 +11,7 @@
 [build-system]
 requires = [
     "setuptools_scm[toml]>=6.2",
-    "scikit-build-core[pyproject]",
+    "scikit-build-core[pyproject]==0.5.1",
```
Comment (zerothi) [inline pyproject.toml:14]: please leave this out... It builds fine with later versions, just not editable installs :)

File: src/sisl/geom/flat.py [hunk 2]
```diff
@@ -73,3 +73,107 @@ def graphene(bond=1.42, atoms=None, orthogonal=False):
...
+    def _minimal_op(shells):
+
+        # The function is based on the horizontal lines of the hexagon,
+        # which are made of a pair of atoms.
+        # For each shell, we first need to complete the incomplete horizontal
+        # lines of the previous shell, and then branch them up and down to create
+        # the next horizontal lines.
+
+        # Displacement from the end of one horizontal pair to the beggining of the next
+        branch_displ_x = bond * np.cos(np.radians(60))
```
Comment (zerothi) [inline src/sisl/geom/flat.py:105]: `r = pi/3`, `np.cos(pi/3) == 0.5`w
`np.sin(pi/3) == 3 ** 0.5 / 2`

**Review by zerothi — CHANGES_REQUESTED** [review]
great!

If adding 2 atoms, would it then alternate in each hexagon? That would be my ideal usecase, no?

Sorry, haven't tested, but looks good!

**Comment by pfebrer** [comment]
> Ok, but anyways, I think this is fitting

Good :+1: I will then finish the PR some time this week probably.

To do the section stuff, it would be useful that the `graphene_flake` stored metadata on the `Geometry`, like `edge_axes` (the axes perpendicular to edges), `edge_R` (how far is the edge from the center) or `edge_W` (the number of C in the edge). But this requires the transformable attributes that we've discussed many times already :sweat_smile:

File: pyproject.toml (hunk 1 above)
Comment (pfebrer) [reply pyproject.toml:14]: Yes, I will remove it :+1:

File: src/sisl/geom/flat.py (hunk 2 above)
Comment (pfebrer) [reply src/sisl/geom/flat.py:105]: Ok!

**Comment by pfebrer** [comment]
> If adding 2 atoms, would it then alternate in each hexagon? That would be my ideal usecase, no?

See the picture in the initial PR comment, is this how you expect it? Is there any other way?

**Comment by zerothi** [comment]
> > If adding 2 atoms, would it then alternate in each hexagon? That would be my ideal usecase, no?
> 
> See the picture in the initial PR comment, is this how you expect it? Is there any other way?

Sorry, too fast ;) Beautiful! :)

**Comment by pfebrer** [comment]
Ok, this PR is ready :+1:

File: src/sisl/geom/flat.py [hunk 3]
```diff
@@ -73,3 +73,107 @@ def graphene(bond=1.42, atoms=None, orthogonal=False):
...
+    Parameters
+    ----------
+    shells:
+        Number of shells in the flake. 0 means a single hexagon, and subsequent
+        shells add hexagon units surrounding the previous shell.
+    bond:
+        bond length between atoms (*not* lattice constant)
+    atoms:
+        the atom (or atoms) that the honeycomb lattice consists of
+    vaccuum:
```
Comment (zerothi) [inline src/sisl/geom/flat.py:89]: Spell, and all the way through

Comment (zerothi) [reply src/sisl/geom/flat.py:89]: Sorry, was on the phone, `vacuum`, and everywhere else :)

Comment (pfebrer) [reply src/sisl/geom/flat.py]: Fixed it :+1:

**Comment by pfebrer** [comment]
Should I set the boundary conditions to `open`?

**Comment by zerothi** [comment]
> Should I set the boundary conditions to `open`?

All you need to do is to call `geometry_define_nsc(geom, [False, False, False])` then return `geom`.

This will set the boundary conditions correctly, to UNKNOWN.

**Comment by pfebrer** [comment]
Why do you say the correct bc is UNKNOWN and not OPEN? It is a 0D flake.

**Comment by zerothi** [comment]
> Why do you say the correct bc is UNKNOWN and not OPEN? It is a 0D flake.

For this it would be better to use `Dirichlet`, `OPEN` is a reference to NEGF, perhaps `OPEN` should be bulk, or something. Typically OPEN is used in the nomenclature of NEGF BC.

**Comment by pfebrer** [comment]
Ok, it should be fine now :+1:

File: src/sisl/geom/flat.py [hunk 4]
```diff
@@ -73,3 +73,112 @@ def graphene(bond=1.42, atoms=None, orthogonal=False):
...
+    max_x = np.max(geom.xyz[:,0])
+    geom.cell[0,0] = max_x * 2 + vacuum
+    geom.cell[1,1] = max_x * 2 + vacuum
+    geom.cell[2,2] = 20.
+
+    # Center the flake
+    geom = geom.translate(geom.center(what="cell"))
+
+    # Set boundary conditions
+    geom.lattice.set_boundary_condition(a="Unknown", b="Unknown", c="Unknown")
```
Comment (zerothi) [inline src/sisl/geom/flat.py:158]: No, call `geometry_set_nsc(geom, [False, False, False])`, `geometry_set_nsc` does other things as well. :)

File: src/sisl/geom/flat.py [hunk 5]
```diff
@@ -73,3 +73,112 @@ def graphene(bond=1.42, atoms=None, orthogonal=False):
     if atoms is None:
         atoms = Atom(Z=6, R=bond * 1.01)
     return honeycomb(bond, atoms, orthogonal)
+
+def honeycomb_flake(shells: int, bond: float, atoms, vacuum: float = 20.) -> Geometry:
```
Comment (zerothi) [inline src/sisl/geom/flat.py:77]: missing `@set_module("sisl.geom")`

File: src/sisl/geom/flat.py [hunk 6]
```diff
@@ -73,3 +73,112 @@ def graphene(bond=1.42, atoms=None, orthogonal=False):
...
+
+    # Center the flake
+    geom = geom.translate(geom.center(what="cell"))
+
+    # Set boundary conditions
+    geom.lattice.set_boundary_condition(a="Unknown", b="Unknown", c="Unknown")
+
+    return geom
+
+def graphene_flake(shells: int, bond: float = 1.42, atoms=None, vacuum: float = 20.) -> Geometry:
```
Comment (zerothi) [inline src/sisl/geom/flat.py:162]: missing `@set_module("sisl.geom")`

File: src/sisl/geom/flat.py (hunk 4 above)
Comment (pfebrer) [reply src/sisl/geom/flat.py:158]: The only thing that it does is to set nsc to `[1, 1, 1]`, no? But to do that it calls first `geometry.optimize_nsc` which is a waste of time in this case...

Comment (zerothi) [reply src/sisl/geom/flat.py:159]: The point is that it we have a single reference function that sets up everything. All `geom` codes call this one. And so it should also.

Comment (pfebrer) [reply src/sisl/geom/flat.py:159]: Hmm ok but then I think `geometry_set_nsc` should check if all directions are not periodic and don't call `optimize_nsc` then

Comment (zerothi) [reply src/sisl/geom/flat.py:159]: True, I'll do that afterwards, call it and I'll get it in.

Comment (pfebrer) [reply src/sisl/geom/flat.py]: Done

**Review by zerothi — APPROVED** [review]

**Pull request merged**